lib/plug/telemetry.ex

```diff
@@ -0,0 +1,31 @@
+defmodule Plug.Telemetry do
+  @moduledoc """
+  A plug to instrument the pipeline with `:telemetry` events.
+  """
+
+  @behaviour Plug
+
+  @impl true
+  def init(_), do: []
+
+  @impl true
+  def call(conn, _opts) do
+    time = System.monotonic_time()
+    :telemetry.execute([:plug, :call, :start], %{}, %{conn: conn})
+
+    conn
+    |> Plug.Conn.put_private(:plug_telemetry_start_time, time)
+    |> Plug.Conn.register_before_send(&emit_stop_event/1)
```

Comment:
Instead of storing the start time in private, you can pass it it in the closure:

    Plug.Conn.register_before_send(conn, &emit_stop_event(&1, time))